src/liballoc/collections/btree/set.rs

```diff
@@ -109,6 +109,68 @@ pub struct Range<'a, T: 'a> {
     iter: btree_map::Range<'a, T, ()>,
 }
 
+/// Core of SymmetricDifference and Union.
+/// More efficient than btree.map.MergeIter,
+/// and crucially for SymmetricDifference, next() reports on both sides.
+#[derive(Clone)]
+struct MergeIter<I>
+where
+    I: Iterator,
+    I::Item: Copy,
+{
+    a: I,
+    b: I,
+    peeked: Option<MergeIterPeeked<I>>,
+}
+#[derive(Copy, Clone, Debug)]
+enum MergeIterPeeked<I>
+where
+    I: Iterator,
+{
+    A(Option<I::Item>),
+    B(Option<I::Item>),
+}
+impl<I> MergeIter<I>
+where
+    I: ExactSizeIterator + FusedIterator,
+    I::Item: Copy + Ord,
+{
+    fn new(a: I, b: I) -> Self {
+        MergeIter { a, b, peeked: None }
+    }
+
+    fn next(&mut self) -> (Option<I::Item>, Option<I::Item>) {
+        let mut a_next = match self.peeked {
+            Some(MergeIterPeeked::A(next)) => next,
+            _ => self.a.next(),
+        };
+        let mut b_next = match self.peeked {
+            Some(MergeIterPeeked::B(next)) => next,
+            _ => self.b.next(),
+        };
```

Comment:
By the way, if I define both (a_next, b_next) with a single match expression (which looks more elegant to me), the performance gain vanishes completely.

Reply:
That's unfortunate, it feels like that kind of detail has been there for a loong time (since early Rust 1.x), that tuple match has inferior codegen.